nizarmah/urban-tree
# Implement frontend client to update streamed snapshots live

### Description

This change implements a typescript react client using Vite. The client retrieves information from the server-side event source and renders them in a tree view.

The rendering in a tree view was a little bit difficult and caused an entire rollercoaster.

Initially, the problem was approached with states and prop drilling. The root node of the tree containing all files was stored initially in a state hook. Each level was rendered recursively by passing the node and its children.

Although this worked great for the initial content, it did not work when it came to modifications to the tree. The modifications done were being done to children of the root node. Accordingly, the root node wasn't actually being affected by the changes. So, the rendered content wouldn't change, even though the tree was changing.

In order to target this in a different manner, I attempted to use reduce hooks and structuring the data in a different manner. The one that worked wonders was structuring the data in a different manner.

Accordingly, I stopped trying to use "nodes" in the client, because they aren't needed. Instead, I relied on the "memoized" nodes which is a hashmap of the node's path with the list of children it has (`hashmap<nodePath, nodeChildren[]>`).

 In order to do that effectively, I had to learn all about react hooks in order to understand how to create my own in the manner required.

Accordingly, the client side now has two different hooks and a static layer containing the memoized tree. One of the hooks is responsible for opening only 1 single connection with the server-side event source and building the memoized tree. The other hook is responsible for notifying the components about changes in its child nodes.

By approaching things in this manner, only the affected branch (parent and direct children) are re-rendered.

### Testing

* Run the nodejs server with some directories passed as arguments

```sh
npm run deploy <dir1> <dir2> ...
```

* Run the client server

```sh
npm run dev
```

* Make changes to the directory and make sure no errors occur while the tree updates

### Notes

* Tests are missing from most of my pull requests, but to be honest, for the short time I'll spend on this project, I'm omitting tests.
* Things need to be documented properly, especially the solutions alongside how to run the client. I'll do that in a future commit in this PR.
